File: solutions/Medium/Find_Minimum_in_Rotated_Sorted_Array_0153.js
/**
 * Function: findMin
 * Description: Given a rotated sorted array of unique elements, this function finds
 * the minimum element in O(log n) time using binary search.
 *
 * @param {number[]} nums - A rotated sorted array of unique integers.
 * @returns {number} - The minimum element in the array.
 */
const findMin = function(nums) {
    let left = 0, right = nums.length - 1;

    while (left < right) {
        let mid = Math.floor((left + right) / 2);

        // If mid element is greater than the rightmost element, min must be in the right half
        if (nums[mid] > nums[right]) {
            left = mid + 1;
        } else {
            // If mid element is smaller than the rightmost, the minimum could be mid or to the left
            right = mid;
        }
    }

    return nums[left]; // Left now points to the minimum element
};

// Example Test Cases
if (require.main === module) {
    console.log("Test 1: Basic rotated array");
    console.log(findMin([3,4,5,1,2]));
    // Expected output: 1

    console.log("Test 2: Larger rotated array");
    console.log(findMin([4,5,6,7,0,1,2]));
    // Expected output: 0

    console.log("Test 3: Already sorted array (no rotation)");
    console.log(findMin([11,13,15,17]));
    // Expected output: 11

    console.log("Test 4: Rotated at one element");
    console.log(findMin([2,3,4,5,6,7,8,1]));
    // Expected output: 1

    console.log("Test 5: Two-element rotated array");
    console.log(findMin([2,1]));
    // Expected output: 1

    console.log("Test 6: Single-element array");
    console.log(findMin([7]));
    // Expected output: 7
}

/*
Time Complexity: O(log n) - Binary search halves the search space at each step.
Space Complexity: O(1) - Only uses constant extra space.
*/

export default findMin;

File: solutions/Medium/Find_Minimum_in_Rotated_Sorted_Array_0153.py
class Solution(object):
    def findMin(self, nums):
        """
        :type nums: List[int]
        :rtype: int

        Finds the minimum element in the rotated sorted array nums 
        using O(log n) binary search.
        """
        left, right = 0, len(nums) - 1

        while left < right:
            mid = (left + right) // 2
            # If mid element is greater than right, pivot is in the right half
            if nums[mid] > nums[right]:
                left = mid + 1
            else:
                right = mid

        return nums[left]


# Example usage:
if __name__ == "__main__":
    solution = Solution()

    # Example 1
    nums1 = [3, 4, 5, 1, 2]
    # Output: 1
    print(solution.findMin(nums1))

    # Example 2
    nums2 = [4, 5, 6, 7, 0, 1, 2]
    # Output: 0
    print(solution.findMin(nums2))

    # Example 3
    nums3 = [11, 13, 15, 17]
    # Output: 11
    print(solution.findMin(nums3))

    # Example 4
    nums4 = [1]
    # Output: 1
    print(solution.findMin(nums4))

    print("All test cases passed!")

# Complexity analysis
# Time complexity: O(log n), where n is the length of the input list 'nums'.
# Space complexity: O(1), since we are using only a constant amount of space.
